File: 142/fast.f90


program main
    implicit none
    integer N,S,X,Y,Z,Q
    integer i,j
    integer :: int_arr
    logical,allocatable :: arr(:)

    read *,N,S,X,Y,Z
    call mk_int_arr
    read *,Q
    do i = 1, Q
        call modify_arr
    end do

    do i = 1, N
        if ( arr(i) .eqv. .true. ) then
            write (*,fmt='(a)', advance='no'),"O"
        else
            write (*,fmt='(a)', advance='no'),"E"
        end if
    end do
    write (*,fmt='(a)') ""

contains
subroutine mk_int_arr
    implicit none
    integer i,prev
    logical mod2
    allocate(arr(N))
    prev = S
    arr(1) = mod(prev, 2) == 1
    do i = 1, N-1
        !write(*,*) prev
        prev = mod(X * prev + Y,Z)
        mod2 = mod(prev, 2) == 1
        arr(i+1) = mod2
    end do
    
end subroutine

subroutine modify_arr
    implicit none
    integer CPin1,CPin2,CPout1,CPout2
    integer i,j,length
    logical,allocatable :: mini_arr(:)
    read *,CPin1,CPin2,CPout1,CPout2
    length=CPin2-CPin1+1
    allocate(mini_arr(length))
    do i = 1, length
        mini_arr(i)=arr(CPin1+i-1)
    end do
    do j = 1, length
        arr(j+CPout1-1)=xor(arr(j+CPout1-1),mini_arr(j))
    end do
    !write (*,*) arr
end subroutine


end program  main

File: 142/fast.c
#include <stdio.h>
int N,S,X,Y,Z;

void mk_arr_int (int arr[N]){
    arr[0]=S;
    int prev = S;
    for(int i=1;N>i;i++){
        prev = (X * prev + Y) % Z;
        arr[i] = prev;
    }
}

void mk_arr(int arr[N],int bool_arr[N]){
    for(int i=0;N>i;i++){
        if (arr[i]%2 == 0){
            bool_arr[i] = 0;
        }
        else{
            bool_arr[i] = 1;
        }
    }
}

void mask_and_xor(int arr[N],int i,int j,int m,int n){
    int length = n-m;
    int mini_arr[length];
    for(int c=i-1;c<j;c++){
        mini_arr[c-i] = arr[c]; 
    }
    for(int c=m-1;c<n;c++){
        arr[c] = mini_arr[c-m] ^ arr[c];
    }
}

int main(){
    
    scanf("%d %d %d %d %d",&N,&S,&X,&Y,&Z);
    int arr[N];
    int bool_arr[N];
    int Q;
    int I,J,K,L;
    scanf("%d",&Q);
    mk_arr_int(arr);
    mk_arr(arr,bool_arr);
    
    for(int i=0;i<Q;i++){
        scanf("%d %d %d %d",&I,&J,&K,&L);
        mask_and_xor(bool_arr,I,J,K,L);
    }


    for(int i=0;i<N;i++){
        if (bool_arr[i]==1){
            printf("%s","O");
        }
        else{
            printf("%s","E");
        }
    }
    return 0;
}